iceoryx_utils/include/iceoryx_utils/internal/cxx/list.inl

```diff
@@ -0,0 +1,713 @@
+}
+
+template <typename T, uint64_t Capacity>
+inline const typename list<T, Capacity>::size_type& list<T, Capacity>::getPrevIdx(const const_iterator& iter) const
+    noexcept
+{
+    return getPrevIdx(iter.m_iterListNodeIdx);
+}
+template <typename T, uint64_t Capacity>
+inline typename list<T, Capacity>::size_type& list<T, Capacity>::getPrevIdx(const const_iterator& iter) noexcept
+{
+    return getPrevIdx(iter.m_iterListNodeIdx);
+}
+
+template <typename T, uint64_t Capacity>
+inline const typename list<T, Capacity>::size_type& list<T, Capacity>::getNextIdx(const const_iterator& iter) const
+    noexcept
+{
+    return getNextIdx(iter.m_iterListNodeIdx);
+}
+template <typename T, uint64_t Capacity>
+inline typename list<T, Capacity>::size_type& list<T, Capacity>::getNextIdx(const const_iterator& iter) noexcept
+{
+    return getNextIdx(iter.m_iterListNodeIdx);
+}
+
+
+template <typename T, uint64_t Capacity>
+inline void list<T, Capacity>::setPrevIdx(const size_type idx, const size_type prevIdx) noexcept
+{
+    m_links[idx].prevIdx = prevIdx;
+}
+template <typename T, uint64_t Capacity>
+inline void list<T, Capacity>::setNextIdx(const size_type idx, const size_type nextIdx) noexcept
+{
+    m_links[idx].nextIdx = nextIdx;
+}
+
+
+template <typename T, uint64_t Capacity>
+inline const T* list<T, Capacity>::getDataPtrFromIdx(const size_type idx) const noexcept
+{
+    if (handleInvalidElement(idx))
+    {
+        // errror handling in handleInvalidElement() call
+        return nullptr;
+    }
+
+    return &(reinterpret_cast<const T*>(&m_data)[idx]);
+}
+
+template <typename T, uint64_t Capacity>
+inline T* list<T, Capacity>::getDataPtrFromIdx(const size_type idx) noexcept
+{
+    return const_cast<T*>(static_cast<const list<T, Capacity>*>(this)->getDataPtrFromIdx(idx));
+}
+
+/*************************/
+// failure handling / messaging
+
+template <typename T, uint64_t Capacity>
+inline bool list<T, Capacity>::isValidElementIdx(const size_type idx) const noexcept
+{
+    return (idx < Capacity) && (getPrevIdx(idx) < INVALID_INDEX);
+}
+
+template <typename T, uint64_t Capacity>
+inline bool list<T, Capacity>::handleInvalidElement(const size_type idx) const noexcept
+{
+    // freeList / invalid elements will have the prevIdx set to INVALID_INDEX
+    if (isValidElementIdx(idx))
+    {
+        return false;
+    }
+    else
+    {
+        errorMessage(__PRETTY_FUNCTION__, " malformed index ");
+        std::terminate();
+
+        return true;
+    }
+}
+
+// iterator validition on iterator operations or iterators are function arguments
+// not coherent-save, as not each single iterator dereferencing operation is checked for validity.
+template <typename T, uint64_t Capacity>
+inline bool list<T, Capacity>::handleInvalidIterator(const const_iterator& iter) const noexcept
+{
+    // freeList / invalid elements will have the prevIdx set to INVALID_INDEX
+    if ((getPrevIdx(iter) < INVALID_INDEX) && (getNextIdx(iter) < INVALID_INDEX)
```

Comment:
Works, due to the specific value of INVALID_INDEX being Capacity+1. Note that we cannot and should not check for inequality here as it is implemented.

Changing this would lead to logic errors but this is ok for intricate data structures (implementation details are somewhat rigid, i.e. hard to change, but interface is not).

One thing that does not necessarily needs to be changed is separating the valid check from the handling, i.e. having a method such as isValid(iterator) and a handleInvalidIterator method. As they are always used together right now this is not needed. Should be done when it becomes apparent that valid checks are used on their own.